### ISSUE TITLE
Raise the voice quality

### ISSUE THREAD

**Issue by Leystryku** [Facepunch/garrysmod-requests#479]
Yo,
May you please change the voice sample rate from 11025 to 32000 ( #define VOICE_OUTPUT_SAMPLE_RATE			11025	// Sample rate that we feed to the mixer. | engine/audio/public/voice.h ) and default sv_use_steam_voice  to 0?
Thanks.

**Comment by Exho1** [comment]
+1 for not sounding like I'm speaking through a toaster

**Comment by EthanTheGreat** [comment]
+1 for not sounding like I'm stuck inside the toaster.

**Comment by polivilas** [comment]
+1 for making my dubstep HD

**Comment by YourStalker** [comment]
+1 for toaster dubstep

**Comment by robotboy655** [comment]
Stop with the "+1" spam, it is not helping anyone.

**Comment by UltimateByte** [comment]
I'm just gonna try to ask the good questions here as audio is my main job.
First, let's clarify something : As it's apparently lossy compressed audio, sample rate isn't enough to quantify audio quality, you have to tell about the bitrate too. Of course, sample rate affects the maximum frequency you'll be able to hear. A very good ear can hear from 20hz to around 22000hz (22Khz). The max audio frequency rendered is half the sample rate. so in order to generate an audio signal up to 22Khz, you need a sample rate of at least 44Khz. Audio CD standard sample rate is 44.1Khz. Isn't that perfect ? (video is 48Khz but whatever, fck video)
 
Just wondering : would this really be hard for a server to handle 44,1Khz ? Would this make a noticeable difference ? Even with 20 people speaking, it seems like it could handle it just as a TeamSpeak server does : without using CPU and with negligible bandwidth, for a server... 
BTW, i won't troll too much here, but if only that was bound to another CPU thread.. that wouldn't even be a question.

What's the codec GMod uses already ? Speex ?
Does GMod have a way to change it or is it bound to source engine ?

Maybe something a little bit better (if not already available) like opus voice used in TeamSpeak would be great already. http://www.opus-codec.org/
It supports up to 48Khz and from 6 to 510kb/s (more than MP3 with a better latency and lower CPU usage for compressing).
Just in case, don't even think about an MP3 voice stream, it would use lot of resources and cause more latency.

It could be presented as an autistic cvar (me like dat) to make it easier for server makers : 
sv_voicequality 0 to 10
0 would be the legacy codec with default settings
1 would be Opus Voice at lowest bitrate and samplerate (for example 16Khz 20kb/s -2,5KB/s/client-), and 10 at the max (reasonable) bitrate and samplerate (for example 44Khz 128Kbits/s -16KB/s/client-). Of course, some tests would be required to find interesting values, and to check if any sample rate can be used with any bitrate, and also to test performance (brandwidth and CPU).

Other way that would need documentation and would be a pain for server owners :
- sv_voicecodec X (i guess i already saw that cvar, not sure if in GMod or CSS)
- sv_voiceratemax X (in Kb/s please, not b/s)

Once it's done, why not go further and add Opus Music, to be used through an API so the server can stream audio to users ?

**Comment by Leystryku** [comment]
The server does nothing with the audio except resend it to all the other clients, the same way the client who voicechatted sent it to the server.
Source Engine's current sample format is mono 16 bit pcm.
iirc, The speex codec included with sourceengine is narrowband so it supports up to 42,4 kbps.
It changes depending on the mode you set. Source Engine seems to use mode 6 which equals ~20kbps. The dll included with the game also allows you to change it to 8, which would be a bit better, by changing the quality passed to the initialization for the src engine codec dll to 5 ( it currently uses 4 ). sv_voicecodec exists, it is in gmod too but it was abused to fuck with clients, so they kiled it. 

Changing to Opus ( which requires them to make a new codec dll themselfs, compatible with the current system ) and changing the voice chatting sys to be able to use these convars  ( sv_voicebitrate, sv_voicesamplerate ) is way more work, and considering this request has been dying since a month the chances of that happening is even lower.

**Comment by UltimateByte** [comment]
Thanks for informations about the current Speex codec.
Just one small correction : If i remember well, we call PCM uncompressed audio with basic format like WAV or AIFF, so if i'm right, Speex would not be PCM.

Just one more question as you seem to know it well : does the command to change quality (sv_voicequality ?) also changes the sample rate, or just the bitrate ?
As the author of this request said, it could change samplerate which i guess would be pretty effortless to implement compared to what i said..

I think my proposal s more of a thing that could be integrated to SRCDS servers at the beginning, it should not be up to GMod team but to Valve's team.

**Comment by bmwalters** [comment]
TF2 just got CELT.
http://www.teamfortress.com/post.php?id=25766

**Comment by JShorthouse** [comment]
This needs to happen, I don't know why this issue has been sitting here for 2 years because the ingame voice quality is honestly awful by modern standards. CS:GO has a better codec / higher bitrate even though it is a game where low latency is far more important so I no reason why gmod can't do the same.

@UltimateByte 's suggestion sounds great, if that's too much work for the developers then I see no reason why they can't take 5 seconds to change a variable like the original post was suggesting.

**Comment by LuaTenshi** [comment]
Why not use [Opus](https://opus-codec.org/)? Even the Speex website just says to use [Opus](https://opus-codec.org/)... https://speex.org/

**Comment by Datamats** [comment]
They just added CELT to the source2013 sdk, css dod and hl:dm

http://store.steampowered.com/news/?appids=243750

**Comment by TheBlueTroll** [comment]
While i would also like to see this happen, Facepunch more then likely have moved on to the new Gmod, S&box. as such i doubt they will do anything except fix bugs in gmod. its pritty sad that we will be stuck with toaster voice, but theres not much that would be done.

**Comment by Frictione** [comment]
I do not understand how this issue still persists. 
TF2 just got Opus codec.
http://www.teamfortress.com/jungleinferno/notes.php
To quote:
"The default codec now defaults to steam voice
Steam voice is now using the opus codec at a higher bitrate than the default CELT codec."

I believe it is not because of the complexity but due to lack of importance given to this issue.

Why doesnt GMOD use steam voice alternatively?

**Comment by TheVingard** [comment]
Please follow through with this suggestion, a codec update has been needed for a long time.

**Comment by djosepaci** [comment]
PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE PLEASE

**Comment by 54** [comment]
It would be nice to see Opus included.

**Comment by ThePotatoPony** [comment]
could you please solve the voice quality issues, its really straining on communities when the voice quality is kinda low forcing people to get a teamspeak server to try and fix the issue themselves. that and the other voice bugs where people randomly can't hear anything

**Comment by ExtReMLapin** [comment]
dumb and silly bump but since speex is harcoded and as far as i understood it's using the voice code of source 2007 (the errors message are in source 2007, not in source 2013), it would be nice to update it to use opus

**Comment by Heldex** [comment]
add

**Comment by midlow** [comment]
yes

**Comment by Kialoa** [comment]
4 years after and there is still old codec

**Comment by UltimateByte** [comment]
Happy birthday to Garry's Mod old codec :)

**Comment by VaasKahnGrim** [comment]
its still better then Discord audio atleast :)

**Comment by UltimateByte** [comment]
Congrats for implementing this.
No congrats for the delay, though. Almost 5 years.

**Comment by ExtReMLapin** [comment]
> Congrats for implementing this.
> No congrats for the delay, though. Almost 5 years.

what an attitude

**Comment by Kialoa** [comment]
> Congrats for implementing this.
> No congrats for the delay, though. Almost 5 years.

the thing is they always had it, it was code capped! can you believe that

**Comment by JShorthouse** [comment]
> what an attitude

I'm thankful that this is finally here but I don't think it's unreasonable at all to be annoyed that it took them 5 whole years to change a single variable. This wasn't some minor issue either, it was a major complaint that all players had.

**Comment by UltimateByte** [comment]
> what an attitude

Problem?
We're talking about Facepunch stuff, I assume it is normal to react with some kind of facepalm or facepucnh to the way issues are handled. 5 years to use a native Source function, sounds like a hell of a lot to me.


> the thing is they always had it, it was code capped! can you believe that

Was there a limitation that prevented Facepunch from implementing this that I was unaware of? If so, then my bad.


> I'm thankful that this is finally here but I don't think it's unreasonable at all to be annoyed that it took them 5 whole years to change a single variable. This wasn't some minor issue either, it was a major complaint that all players had.

Spot on, that's exactly my point. Should I say "Thank you" for finally implementing something we asked 5 years ago, or complain about it because it took so long? I'm quite unsure here. Anyhow, if we keep posting on this, the issue will soon be locked so I don't think we will have a chance to discuss that anyway. :)

**Comment by robotboy655** [comment]
Again, do you want me to pointlessly comment on every issue so you don't "feel ignored" or do you want me to actually solve the issues?
And again, there's only 1 of me and hundreds of your issues. I can't solve everything instantly.
What messages you want upon closing the issue? The issue has been addressed. Would a comment saying "Fixed" change anything at all?
Also this is not customer support. It's a bug tracker.

**Comment by ExtReMLapin** [comment]
>No apologize for the delay


lmao

**Comment by UltimateByte** [comment]
> Again, do you want me to pointlessly comment on every issue so you don't "feel ignored" or do you want me to actually solve the issues?
> And again, there's only 1 of me and hundreds of your issues. I can't solve everything instantly.

I don't understand, are you working for free and are you alone working on this? I hope not.

> What messages you want upon closing the issue? The issue has been addressed. Would a comment saying "Fixed" change anything at all?

People usually appreciate human behaviors, words, explanations, excuses... They also can feel compassion rather than anger if they get why it is difficult to get the work done.

If you need a template, it can look like that:
"Sorry for the delay, I finally had time to implement this, it took so long because <insert reason here> but it's there! To use it, just <insert how to use here>. Hope you appreciate the work. Cheers!"

> Also this is not customer support. It's a bug tracker.

Well, call it and consider it as you wish, it is still allows actual customers (people who payed a product) to get fixes and improvements. In my book, after doing sysadmin & customer support for 3 years, people who bought something also have right to get support to actually enjoy and get the full potential of what they bought.

**Comment by robertdrakedennis** [comment]
I wish I could feel this conceited on a daily basis

**Comment by ExtReMLapin** [comment]
When you don't understand why the developer personally fixing your problem (which is not even a problem but a request) isn't acting like a L1 support.

**Comment by UltimateByte** [comment]
You're reacting as if Facepunch was a crew of two amateurs working for free. Aren't they an actual company which should have professional standards? Well, maybe not, and I should then lower my standards with them (which I have actually already done a long time ago for them and many others, since mediocrity is kind of the norm in our society, but I sometimes try to push towards higher standards, in moments of hope).

> I wish I could feel this conceited on a daily basis

If you're talking about robotboy655's way, I believe the way is to try being a robot and to consider everyone as a loss of time; consider any of your actions and interactions with others as a blessing. You should achieve pretty similar results.

If you're talking about mine, then work hard, get good work done on a daily basis and be human and friendly with others. You will get happy customers (and coworkers) and will be able to be proud of what you achieved. I hope everyone working actually experiences this. Then upon talking with people who happen to disappoint their customers for various reasons, the confrontation of your points of view might naturally make you look conceited.

> When you don't understand why the developer personally fixing your problem (which is not even a problem but a request) isn't acting like a L1 support.

What do you mean "personally"? Is he working on his own for free, or does he somehow represent Facepunch, being an actual employee with responsibilities and duties?
When the dev interacts with the public, he shall endorse this role of offering technical support, be human and friendly. I don't care if it's a bug fix or a request, what matters is that people care about the issue. Otherwise, keep the dev in a cave doing its stuff and let other people who actually care about others and know how to act in public as a company do the exchange and communication with customers.

**Comment by thegrb93** [comment]
tl;dr

fyi gmod was a finished product over 10 years ago. It's worked on and continually improved because of the passion they have for it and the community. If this was any other company, they would be charging DLC money for updates. Now shut.

**Comment by ExtReMLapin** [comment]
Imagine calling yourself UltimateByte, openly saying your company is HaiSoftSARL and being such a douche.

**Comment by UltimateByte** [comment]
@thegrb93 
Too bad that you couldn't bear reading something longer than a Twitter post.

All about passion... For a game with 10M to 20M owners. Sorry but I find that hard to believe.
https://steamdb.info/app/4000/graphs/

![gmod chart](https://user-images.githubusercontent.com/8805941/80750533-8d12e580-8b28-11ea-83ce-a167c4913820.PNG)

Of course it's still developed, since people keep buying it, it has to get support to keep the player base and keep selling.
If you're working for free, I'm sorry for you. If not, then act as a company.


@ExtReMLapin 

As far as I know, I'm not currently working for customers of my company here, and I have nothing to be ashamed of neither here or at work. If you have any actual argument, please comment. Otherwise you can GTFO and go back playing, you'll be more useful there.

**Comment by ExtReMLapin** [comment]
>As far as I know, I'm not currently working for customers of my company here, and I have nothing to be ashamed of neither here or at work. If you have any actual argument, please comment. Otherwise you can GTFO and go back playing, you'll be more useful there.

At this point it's not just being a keyboard warrior, you just turned yourself into a keyboard commando

**Comment by UltimateByte** [comment]
@ExtReMLapin Very interesting point of you. You can maybe discuss this philosophy with your dog, he might be interested.

**Comment by thegrb93** [comment]
You're implying being popular means they're obligated to push updates. Also an interesting point of view. Your cat may like to hear as well.

**Comment by UltimateByte** [comment]
It goes along. Don't you see why?

If it is popular, that makes loads of support requests, then if it's properly supported, it gives reasons to stay popular. And if it makes money on top of that, that's one hell more of a reason for a company to want to keep it supported, therefore popular to keep making money, because that's what companies does to pay the bills and expand.

**Comment by thegrb93** [comment]
Considering half the issues here are source-engine issues. Why not complain to valve while you're at it? Imo valve are the real shitlords in this situation.

**Comment by Kefta** [comment]
Hire me Facepunch

**Comment by UltimateByte** [comment]
> Considering half the issues here are source-engine issues (INCLUDING THIS ONE). Why not complain to valve while you're at it? Imo valve are the real shit company in this situation.

How could I know since nobody from Facepunch communicates useful info on these issues?

**Comment by JShorthouse** [comment]
>Considering half the issues here are source-engine issues (INCLUDING THIS ONE)

Yes, this is surely a problem with the source engine when every other source game has had perfectly clear audio for years

**Comment by Kefta** [comment]
Because not even Facepunch knows every detail and limitation of the Source engine - iirc the reason https://github.com/Facepunch/garrysmod-requests/issues/647 hasn't even been attempted yet is because how voice chat is exactly handled and passed thru on the server is still a mystery.

**Comment by ExtReMLapin** [comment]
Imagine being this 30yo dude talking like a 13yo angry kid about communication.

**Comment by UltimateByte** [comment]
> Because not even Facepunch knows every detail and limitation of the Source engine - iirc the reason #647 hasn't even been attempted yet is because how voice chat is exactly handled and passed thru on the server is still a mystery.

Well, in that case, what does it cost to say "we didn't yet find a way to implement this"? Or "we found a lead but are stuck on X detail", that's what Github is supposed to be... Devs helping each others.

Lack of communication is the pain here, people waiting with no idea of what's going on.
Especially when on top of that, rare staff answers are pretty much always salty and unprofessional.
Are devs here even Facepunch employees? Because they're acting as volunteers. And if they are employees, is there actually any leadership in this stupid company to guide towards at least the basic baselines of communication?

> 
> 
> Imagine being this 30yo dude talking like a 13yo angry kid about communication.

The only 13yo here is you, bragging while having 0 argument or anything relevant to say, no idea to develop. Your talk is empty, you're the cancer of the internet. Please, go back to your unemployment and leave adults discuss.

**Comment by Kefta** [comment]
> Lack of communication
> rare staff answers
> salty and unprofessional
You were obviously not around for when Garry was the primary developer. Rubat communicates far more than he ever did.

**Comment by UltimateByte** [comment]
@Kefta Sorry for the length of that comment, but there are some things that need a detailed answer to demonstrate a point.

It is true that I was following from a distance on a forum back then, so I'm mainly talking about what I'm seing now. The past doesn't really matter to evaluate the quality of support here and now.

For communication now... Well, there are changelogs. But from what I've seen before and today on this Git, #647 #479 #1452 #511 , you're pretty much the most active @Kefta and I'm not even sure you're part of the staff... Main devs that I know of remain silent. Nobody is even sorting issues or adding tags. I mean, did anybody even try to find out how Github works to find benefits to using it compared to a simple forum?
Instead, I only see ignored posts, closed issues and free distasteful comments from time to times.

I mean, take this specific issue for example. The only dev comment says "Stop with the "+1" spam, it is not helping anyone." Nice. Very nice.
As the amount of dislikes on this comment demonstrates: Not only is this lacking politeness, it doesn't provide any information regarding the actual issue, any consideration for users, it's just as unprofessional as it can get. That's the kind of things you might say as an independent giving his time for free, but not as a company member (and I sincerely hope they are). There's nothing to be proud of with this kind of behavior and I myself had remorse when answering badly to comments on a git I was volunteer on... And it shall be worst, unacceptable as an employee in a company with proper management.

For having been a very active dev for LinuxGSM project during a few years back in 2014-2017, I can tell that this free & open source project primarily managed by one single man is maintained way more cleanly and efficiently than Facepunch's one.
To give you an idea:
https://github.com/GameServerManagers/LinuxGSM/issues
Just look at tags, at issues opened by the creator himself, and the amount of answers by people who actually try to solve issues there. That is organized. This Facepunch git is not, it's a real mess, a trash can.

If one man can do so well (even though not perfect but still) for a free project, how can a company managing a project for 10M paid users struggle so much to do things decently? I can only think of one thing: poor management. #Facepunch

**Comment by thegrb93** [comment]
They don't use github for issue tracking. This is just a public tracker for community members to report things.

**Comment by UltimateByte** [comment]
> They don't use github for task management. This is just a public tracker for community members to report things.

Then why use Github rather than a stupid forum? Don't act like an Apple fanboy justifying every single mistake they make. Making public tool for bug reporting implies showing some respect to people participating and keeping it at least somewhat organized.

**Comment by thegrb93** [comment]
That's your opinion. Judging from the last few changelogs, it's been working out quite well.

**Comment by thegrb93** [comment]
@robotboy655 mute this thread already

**Comment by ExtReMLapin** [comment]
I think it would be wiser to ban the guy from FP repos as he's not contributing at all.

**Comment by JShorthouse** [comment]
This whole discussion is stupid, but @UltimateByte isn't saying anything wrong.

Just lmao at all the kids in here coming out to provide free defence for a multi-million dollar game studio. Remember who is paying who.

**Comment by thegrb93** [comment]
It's not wrong or right, it's opinion. We're as free to disagree with him as he is to give his opinion. Calling people kids is just as toxic as his shit attitude.

**Comment by UltimateByte** [comment]
What's the problem with opinions as long as they are argumented?

**Comment by thegrb93** [comment]
There isn't, it's the condescending rhetoric and insults.

**Comment by Kefta** [comment]
Please move this discussion to the GMod discord or anywhere else but this repo. This is an issue tracker and watchers are notified for every comment left.

**Comment by Leystryku** [comment]
For the love of christ, may you both please stop posting here?
My mailbox keeps getting flooded by all the notifications.

Both sides are entirely, yes it could've been done quicker but also yes it's a big game with only 2 developers actively working on it.

Either way, arguing here solves nothing.

**Comment by JShorthouse** [comment]
> with only 2 developers actively working on it

Which is entirely the fault of facepunch, and not an excuse. May I remind you that this game has probably made ~$50M gross, we're not talking about some niche indie studio that can't afford staff.

**Comment by thegrb93** [comment]
They are free to do with that money what they want, and for most of their studio, that is develop other games. If you want to apply as a Garry's Mod dev, prove your worth and apply for it.

**Comment by UltimateByte** [comment]
> They are free to do with that money what they want, and for most of their studio, that is develop other games. If you want to apply as a Garry's Mod dev, prove your worth and apply for it.
> 
> Developers don't grow on trees.

We are also free to criticize. Without criticizing, no evolution, no improvements. Shouting on Discord won't improve anything either.

**Comment by robotboy655** [comment]
Yeah, closing this.

@UltimateByte Please accept my apologies if my behavior is not up to your standards. I can only try to be better,
